### PR TITLE
Added the Frostlands

### PR DESCRIPTION
**Pull request by Laugic**
A zone that is currently 1 encounter and 1 elite fight contained within an event.
Adds relics as well.

### CODE REVIEW

File: src/main/java/spireMapOverhaul/zones/frostlands/FrostlandsZone.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            rewardItem.incrementGold(ELITE_EXTRA_GOLD);
+        }
+    }
+
+    @Override
+    public void modifyRewards(ArrayList<RewardItem> rewards) {
+        for (int i = 0; i < rewards.size(); i++) {
+            if (rewards.get(i).type == RewardItem.RewardType.RELIC) {
+                RewardItem rewardItem = new RewardItem();
+                //Note that this reward does not get the normal act-dependent chance of cards being upgraded
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/FrostlandsZone.java:87]: Is there a reason you'r not giving it the normal upgrade logic? We have a utility method for it in RewardModifyingZone for it now

File: src/main/java/spireMapOverhaul/zones/frostlands/FrostlandsZone.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                        r.onPreviewObtainCard(c);
+                    }
+                }
+                rewards.set(i, rewardItem);
+            }
+        }
+    }
+
+    @Override
+    public void postCreateShopCards(ArrayList<AbstractCard> coloredCards, ArrayList<AbstractCard> colorlessCards) {
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/FrostlandsZone.java:100]: I don't think you wanted to copy this?

File: src/main/java/spireMapOverhaul/zones/frostlands/cardmods/FrostbiteModifier.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.Random;
+
+import static spireMapOverhaul.SpireAnniversary6Mod.makeID;
+import static spireMapOverhaul.SpireAnniversary6Mod.modID;
+import static spireMapOverhaul.util.Wiz.atb;
+
+public class FrostbiteModifier extends AbstractCardModifier {
+
+    public static String ID = makeID(FrostbiteModifier.class.getSimpleName());
+    private static final Texture tex = TexLoader.getTexture(modID + "Resources/images/ui/Frostlands/Frostbite.png");
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/cardmods/FrostbiteModifier.java:49]:
```suggestion
    private static final Texture tex = TexLoader.getTexture(SpireAnniversary6Mod.makeUIPath("Frostlands/Frostbite.png"));
```

File: src/main/java/spireMapOverhaul/zones/frostlands/cardmods/FrostbiteModifier.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    @Override
+    public List<TooltipInfo> additionalTooltips(AbstractCard card) {
+        ArrayList<TooltipInfo> tooltips = new ArrayList<>();
+        tooltips.add(new TooltipInfo(BaseMod.getKeywordTitle(makeID("frostbite")), BaseMod.getKeywordDescription(makeID("frostbite"))));
+        return tooltips;
+    }
+
+//    @Override
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/cardmods/FrostbiteModifier.java:79]: Please remove code that you don't need anymore

File: src/main/java/spireMapOverhaul/zones/frostlands/cardmods/FrostbiteModifier.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+//    public void onUpdate(AbstractCard card) {
+//    }
+    @Override
+    public void onDrawn(AbstractCard card) {
+        card.dontTriggerOnUseCard = false;
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/cardmods/FrostbiteModifier.java:84]: Huh?

File: src/main/java/spireMapOverhaul/zones/frostlands/events/SnowmanMafiaEvent.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    public static boolean usedContraption;
+
+
+    private enum CurScreen {
+        INTRO, ACCEPT, DENY, POST_COMBAT, RECOVER, LEAVE;
+    }
+
+    public SnowmanMafiaEvent() {
+        super(NAME, DESCRIPTIONS[0], IMG);
+        imageEventText.setDialogOption(OPTIONS[0]);
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/events/SnowmanMafiaEvent.java:41]: You can add a relic preview here iirc

File: src/main/java/spireMapOverhaul/zones/frostlands/events/SnowmanMafiaEvent.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                return;
+            case LEAVE:
+                openMap();
+                return;
+        }
+        openMap();
+    }
+
+    public void logMetric(String actionTaken) {
+        AbstractEvent.logMetric("Colosseum", actionTaken);
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/events/SnowmanMafiaEvent.java:131]: I guess, this is kinda an analogue?

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Cole.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        move = AbstractDungeon.monsterRng.random(1, 2);
+    }
+
+    @Override
+    public void usePreBattleAction() {
+        AbstractDungeon.actionManager.addToBottom(new ApplyPowerAction(this, this, new LeavePower(this)));
+    }
+
+    @Override
+    public void takeTurn() {
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/monsters/Cole.java:59]: No harder moveset fr a17/18?

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Spruce.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        type = EnemyType.NORMAL;
+        if (AbstractDungeon.ascensionLevel >= 7)
+        {
+            setHp(HP+8, HP_MAX+8);
+            damage.add(new DamageInfo(this, A7_DMG));
+        }
+        else
+        {
+            setHp(HP, HP_MAX);
+            damage.add(new DamageInfo(this, DMG));
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/monsters/Spruce.java:50]: HP and damage modification doN#t come from the smae asc

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Steward.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+                IMG, x, y);
+        currentX = x;
+        currentY = y;
+        currentHB = ((float) (HP + 64) /(float)((AbstractDungeon.ascensionLevel >= 7)?(A7_HP + 64):(HP + 64)));
+        loadAnimation(modID + "Resources/images/monsters/Frostlands/Steward.atlas", modID + "Resources/images/monsters/Frostlands/Steward.json",
+                 1 / ((float) (HP + 64) /(float)((AbstractDungeon.ascensionLevel >= 7)?(A7_HP + 64):(HP + 64))));
+        state.setAnimation(0, "idle", true);
+        firstMove = false;
+        spawned = true;
+        type = EnemyType.NORMAL;
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/monsters/Steward.java:78]: btw, isn't this an elite fight?

File: src/main/java/spireMapOverhaul/zones/frostlands/patches/FrostPatches.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+import com.megacrit.cardcrawl.cards.AbstractCard;
+import com.megacrit.cardcrawl.characters.AbstractPlayer;
+import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
+import com.megacrit.cardcrawl.monsters.AbstractMonster;
+import javassist.CannotCompileException;
+import javassist.CtBehavior;
+import spireMapOverhaul.zones.frostlands.cardmods.FrostbiteModifier;
+import spireMapOverhaul.zones.manasurge.modifiers.common.negative.FeebleModifier;
+
+public class FrostPatches {
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/patches/FrostPatches.java:17]: What's happening here?

File: src/main/java/spireMapOverhaul/zones/frostlands/powers/ChillPower.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+import static spireMapOverhaul.SpireAnniversary6Mod.makeID;
+
+public class ChillPower  extends AbstractSMOPower {
+    public static String POWER_ID = makeID(ChillPower.class.getSimpleName());
+    public static PowerStrings strings = CardCrawlGame.languagePack.getPowerStrings(POWER_ID);
+    public static String[] DESCRIPTIONS = strings.DESCRIPTIONS;
+
+    public ChillPower(AbstractCreature owner, int amount) {
+        super(POWER_ID, strings.NAME, "Frostlands", AbstractPower.PowerType.DEBUFF, true, owner, amount);
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/powers/ChillPower.java:21]: Why use the string literal here? Also, this isn't really turn based

File: src/main/resources/anniv6Resources/localization/eng/Frostlands/Keywordstrings.json [hunk 13]
```diff
@@ -0,0 +1,10 @@
+[
+  {
+    "ID": "frostbite",
+    "PROPER_NAME": "Frostbite",
+    "NAMES": [
+      "frostbite"
+    ],
+    "DESCRIPTION": "If this card is in your hand at the end of your turn, take 4 damage."
```
Comment (erasels) [inline src/main/resources/anniv6Resources/localization/eng/Frostlands/Keywordstrings.json:8]:
```suggestion
    "DESCRIPTION": "If this card is in your hand at the end of your turn, take #b4 damage."
```

File: src/main/resources/anniv6Resources/localization/eng/Frostlands/Powerstrings.json [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  "${ModID}:LeavePower": {
+    "NAME": "Leave",
+    "DESCRIPTIONS": [
+      "If this creature would be killed in combat, they leave instead."
+    ]
+  },
+  "${ModID}:ChillPower": {
+  "NAME": "Chill",
+  "DESCRIPTIONS": [
+    "Decreases Attack damage by #b",
```
Comment (erasels) [inline src/main/resources/anniv6Resources/localization/eng/Frostlands/Powerstrings.json:20]:
```suggestion
    "Decreases attack damage by #b",
```

File: src/main/resources/anniv6Resources/localization/eng/Frostlands/Relicstrings.json [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    "FLAVOR": "Hypothema's hand-me-down.",
+    "DESCRIPTIONS": [
+      "At the start of your turn, gain #b", " #yVigor."
+    ]
+  },
+  "${ModID}:Frostcoal": {
+    "NAME": "Frostcoal",
+    "FLAVOR": "Chip off the old block.",
+    "DESCRIPTIONS": [
+      "Enemies deal #b", " less damage for the first turn of combat."
```
Comment (erasels) [inline src/main/resources/anniv6Resources/localization/eng/Frostlands/Relicstrings.json:20]:
```suggestion
      "Enemies deal #b", " less damage during the first turn of combat."
```

File: src/main/resources/anniv6Resources/localization/eng/Frostlands/UIstrings.json [hunk 16]
```diff
@@ -0,0 +1,9 @@
+{
+  "${ModID}:Frostlands": {
+    "TEXT": [
+      "Frostlands",
+      "A frosty crew can be found chilling in the Frostlands within the Spire. If you can overcome their challenges, they may have powerful rewards for you.",
```
Comment (erasels) [inline src/main/resources/anniv6Resources/localization/eng/Frostlands/UIstrings.json:5]:
```suggestion
      "A #bfrosty #bcrew can be found #bchilling in the Frostlands within the Spire. If you can overcome their challenges, they may have #gpowerful #grewards for you.",
```

**Review by erasels — CHANGES_REQUESTED** [review]

File: src/main/java/spireMapOverhaul/zones/frostlands/events/SnowmanMafiaEvent.java (hunk 6 above)
Comment (Laugic) [reply src/main/java/spireMapOverhaul/zones/frostlands/events/SnowmanMafiaEvent.java:41]: Added

File: src/main/java/spireMapOverhaul/zones/frostlands/patches/FrostPatches.java (hunk 11 above)
Comment (Laugic) [reply src/main/java/spireMapOverhaul/zones/frostlands/patches/FrostPatches.java:17]: Was for old functionality - removed

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Cole.java (hunk 8 above)
Comment (Laugic) [reply src/main/java/spireMapOverhaul/zones/frostlands/monsters/Cole.java:59]: Done

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Spruce.java (hunk 9 above)
Comment (Laugic) [reply src/main/java/spireMapOverhaul/zones/frostlands/monsters/Spruce.java:50]: Fixed

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Cole.java [hunk 17]
```diff
@@ -36,17 +37,20 @@ public Cole(final float x, final float y) {
...
-        if (AbstractDungeon.ascensionLevel >= 7)
+        type = EnemyType.ELITE;
+        buff = 0;
+        int hp = HP;
+        blk = BLK;
+        if (AbstractDungeon.ascensionLevel >= 8)
+            hp+=10;
+        if (AbstractDungeon.ascensionLevel >= 3)
+            blk += 2;
+        if (AbstractDungeon.ascensionLevel >= 17)
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/monsters/Cole.java:48]:
```suggestion
        if (AbstractDungeon.ascensionLevel >= 18)
```

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Cole.java [hunk 18]
```diff
@@ -36,17 +37,20 @@ public Cole(final float x, final float y) {
...
-        }
-        else
-        {
-            setHp(HP, HP_MAX);
-            blk = BLK;
+            buff++;
+            blk += 2;
         }
+        setHp(hp, hp+10);
         move = AbstractDungeon.monsterRng.random(1, 2);
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/monsters/Cole.java:54]:
```suggestion
        move = AbstractDungeon.aiRng.random(1, 2);
```

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Hypothema.java [hunk 19]
```diff
@@ -38,17 +38,21 @@ public Hypothema(final float x, final float y) {
...
-        if (AbstractDungeon.ascensionLevel >= 7)
+        type = EnemyType.ELITE;
+        buff = 0;
+        int hp = HP;
+        int dmg = DMG;
+        if (AbstractDungeon.ascensionLevel >= 8)
+            hp+=10;
+        if (AbstractDungeon.ascensionLevel >= 3)
+            dmg++;
+        if (AbstractDungeon.ascensionLevel >= 17)
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/monsters/Hypothema.java:49]:
```suggestion
        if (AbstractDungeon.ascensionLevel >= 18)
```

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Hypothema.java [hunk 20]
```diff
@@ -38,17 +38,21 @@ public Hypothema(final float x, final float y) {
...
-        else
-        {
-            setHp(HP, HP_MAX);
-            damage.add(new DamageInfo(this, DMG));
+            buff++;
+            dmg++;
         }
+        setHp(hp, hp+10);
+        damage.add(new DamageInfo(this, dmg));
         move = AbstractDungeon.monsterRng.random(1, 2);
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/monsters/Hypothema.java:56]:
```suggestion
        move = AbstractDungeon.aiRng.random(1, 2);
```

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Spruce.java [hunk 21]
```diff
@@ -38,18 +38,18 @@ public Spruce(final float x, final float y) {
...
+        if (AbstractDungeon.ascensionLevel >= 8)
+            hp+=10;
+        if (AbstractDungeon.ascensionLevel >= 3)
+            dmg++;
+        if (AbstractDungeon.ascensionLevel >= 17)
+            dmg += 2;
+        setHp(hp, hp+10);
+        damage.add(new DamageInfo(this, dmg));
+        damage.add(new DamageInfo(this, (int)(.67f * dmg)));
         move = AbstractDungeon.monsterRng.random(1, 2);
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/monsters/Spruce.java:53]:
```suggestion
        move = AbstractDungeon.aiRng.random(1, 2);
```

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Spruce.java [hunk 22]
```diff
@@ -38,18 +38,18 @@ public Spruce(final float x, final float y) {
...
-        }
-        damage.add(new DamageInfo(this, DMG2));
+        type = EnemyType.ELITE;
+        int hp = HP;
+        int dmg = DMG;
+        if (AbstractDungeon.ascensionLevel >= 8)
+            hp+=10;
+        if (AbstractDungeon.ascensionLevel >= 3)
+            dmg++;
+        if (AbstractDungeon.ascensionLevel >= 17)
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/monsters/Spruce.java:48]:
```suggestion
        if (AbstractDungeon.ascensionLevel >= 18)
```

File: src/main/java/spireMapOverhaul/zones/frostlands/monsters/Spruce.java [hunk 23]
```diff
@@ -69,7 +69,7 @@ public void takeTurn() {
             case 2:
                 Wiz.atb(new AnimateSlowAttackAction(this));
                 Wiz.atb(new DamageAction(AbstractDungeon.player, damage.get(1), AbstractGameAction.AttackEffect.SLASH_HORIZONTAL));
-                Wiz.atb(new GainBlockAction(this, BLK));
+                Wiz.atb(new GainBlockAction(this, damage.get(1).base));
```
Comment (erasels) [inline src/main/java/spireMapOverhaul/zones/frostlands/monsters/Spruce.java:72]: Just to be sure, you wanted this?

**Review by erasels — CHANGES_REQUESTED** [review]

**Pull request merged**